### ISSUE TITLE
Various use-cases for first boot/machine-id are broken

### ISSUE THREAD

**Issue by boucman** [systemd/systemd#14131]
**systemd version the issue has been seen with**
> 243.4

**Used distribution**
 > any (tested on buildroot

Ok so i'll nee advice on this one. 
first a little note on my setup/problematic

In the embedded world we usually deal with a red-only rootfs partition and a read-write /var.
* We can't set /etc/machine-id at build-time becuse all products would have the same machine-id
* So the file backing machine-id must be in /var, with /etc red-only
The common idea is to have a symlink from /etc/machine-id to /var/machine-id but that doesn't really work

# The current behavior
## Rootfs is RO then RW (systemd-remount-fs.service)
* no machine-id : error message at boot, not machine-id; firstboot 
* empty machine-id : transient then comitted ; no firstboot **bug : no first_boot**

## Rootfs is RW from the start
* no machine-id : machine-id created from start ; firstboot
* machine-id is empty : machine-id created from start ; no firstboot

## Rootfs is RO 
* no machine-id : error message at boot, not achine-id; firstboot
* empty machine-id : transient ; no firstboot

## Rootfs is RO with a /var RW (mounted before systemd)
* machine-id is symlink to empty : machine-id is created from start ; no firstboot
* machine-id is symlink to non-existant : machine-id is created from start ; no firstboot (/etc is RO)  **bug: we should run firstboot**

## Rootfs is RO with a /var RW (mounted by systemd)
* machine-id is symlink : error message at boot, not machine-id; firstboot **bug/FR : we should support that use-case**

# Problems we observe
## starting RO then RW is broken : it is not possible to have both first-boot and machine-id working
* if machine-id is empty, you don't have first-boot
* if machine-id is non-existant, the system boots with no machine-id at all.

I am not sure how to fix that... my best guess would have to put a "this is a first boot" marker in /etc/machine-id to trigger the firstboot logic. This would be overwritten by systemd-machine-id-commit.

We don't have the information "/etc will become available" early in the boot, nor the information "this is a first boot" we need one or the other to be available in order to fix this. 

Advice welcomed...

## having an early /var almost works
Both machine-id and ConditionFirstBoot work correctly. when /etc/machine-id is a symlink to a writable /var

The only issue is that systemd-firstboot.service itself checks that /etc is writable instead of 
trusting ConditionFirstBoot, but that's easy to fix. I'll do a separate PR

## Having a late /var is problematic.
Having /etc/machine-id be a dangling symlink to /var/machine-id is broken because you can't use a dangling symlink as a mount-point. So the whole transient machine-id logic won't work

Having it be an empty file doesn't work, because "empty file" already means "transient machine-id"

Creating a file /var/machine-id in the read-only root partition, and overmounting /var might work (would need to test that) but it would break the factory logic

# conclusion

I'm ready to work on that, but I need some advice on where to go from here... 
Thx
Boucman
(related bug #8268)
(related PR #14059)

**Comment by poettering** [comment]
i wonder how much of this is left once #16939 is merged.

**Comment by boucman** [comment]
I'll test once that is merged.

This won't solve the late /var problems though (AFAICT) but you stated on the mailing list that you don't see that use-case as valid (i.e having a completely read-only /etc and somehow making a couple of files R/W via bind-mounts or symlinks)


That's pretty problematic for the embedded use-case where this is a common pattern.

having a read-only rootfs that won't become RW even for common file is needed because the rootfs can be in a PROM (rarely, but not unheard of) or because we are  using an A/B upgrade scheme (and all changes to config files are lost on upgrade)

On embedded systems, most config files are not touched for the whole life of a product but there are usually a couple that need to be tweaked (for example: the name/passwd of the wifi on a consumer product). machine-id is a file that we can't pre-set in the gold-image, but need to survive reboots and upgrades

An aditionnal constraint is that the data partition is usually unformatted on first boot, because that partition can be huge and formating increases the factory time greatly.

There are a lot of workarounds for various config files which are read late enough that /var can be mounted, but that doesn't work for the machine-id. So lots of embedded systems have a machine-id that changes on every boot.

Anyway. I still am completely under water with professional stuff, i'll try to get back to it soon...

**Comment by poettering** [comment]
> having a read-only rootfs that won't become RW even for common file is needed because the rootfs can be in a PROM (rarely, but not unheard of) or because we are using an A/B upgrade scheme (and all changes to config files are lost on upgrade)

There seems to be a logical disconnect in this. You want a ready-only root but you also want to make persistent changes. How can that work? Your way out appears to be to move individual bits out of /etc via various hacks. But then, why make /etc read-only in the first place?

So, I very much sympathize with the embedded case, but I fail to grok the point of immutable root but then also want to make changes. The way I see it the following two options make sense:

1. You actually want zero persistency, no local config, and thus full rootfs immutability: in this mode / and /etc is read-only and the machine ID randomized afresh on every boot.

2. You want persistency, and some local config, and thus writable rootfs — even if the vendor OS data (i.e. binaries) is immutable. In this case what I think makes most sense, and what is the model I think we should all go for is that the rootfs is writable, but /usr is separate from it and immutable. Ideally, the root fs is formatted during the first boot (for example via systemd-repart and mostly empty, i.e /etc should not contain anything that isn't local config. This is the model that systemd-repart, systemd-tmpfiles and systemd-syusers have been written for: clear isolation of what is the deployed "OS image" from what is the local config and state, and the basic structure of /etc, of /var and of / are synthesized on first boot if unpopulated. i.e. in my ideal setup for systems like this an OS would come with with an initial version of /usr (the A copy, possibly configured for verity) and then on first boot we'll create a B copy for it, and we'll format the remaining disk as (possibly encrypted) root fs. 

Hence, my question: why move /etc/machine-id from /etc to /var? Why not just make /etc writable and empty it of everything else? At least systemd's codebase doesn't need anything in /etc really to work, and a good chunk of the rest of the other base OS software is fine with that too.

**Comment by boucman** [comment]
In the embedded use-case we have two types of files in /etc

* locally generated ones with local configuration (for example local network configuration, wifi name and passwd)   There are usually a handefull of those files but we need them to survive updates because we can't recreate them. machine-id is one of those, but it's an extra-complicated one, because it's one of the first files systemd reads and there is very little way to "cheat" with that file.

* system configuration, provided by the vendor to configure the various software we use on the system. Those are provided by the device vendor and usually don't change (but they are not "generic" in that we need to configure these software and can't rely on the default configuration). The only case where they might change, is at upgrade time where new version of the software might need a new configuration or security tweaks in the config file. That is why those file are usually included in the A/B update.

Unfortunately, most software don't handle config file in /usr overlayable via /etc. so the idea of having an empty /etc is not really realistic. 

Making /etc writable conflicts with upgrading it via an A/B logic. An upgrade will loose all cahnges to the partition it upgrades so if /etc/ is upgraded via A/B, all local changes are lost.

Having /etc not be part of the upgrade logic means that we needs to upgrade it somehow. It also means we need to keep two copies of /etc around (for rollback) which is cumbersome and fragile.

Yes... your logic would make sense if everybody respected the rule that you can have two config files, one in /usr and one in /etc, but we are pretty far from that. most software still only read on /etc and the factory system is too late for machine-id anyway.

**Comment by poettering** [comment]
> Unfortunately, most software don't handle config file in /usr overlayable via /etc. so the idea of having an empty /etc is not really realistic.

I am not sure this is really true that much. At least the Fedora images I regularly boot with empty /etc mostly just work.

Also, note that there's tmpfiles.d/ L and C lines, which default to sources in `/usr/share/factory/`. The idea is that for software that insists on reading stuff from /etc you can use that to minimally populate `/etc/` if relevant files are missing, either by linking them from /usr/share/factory/ or by copying things from there.

With systemd we generally try to push people to some solution that in the end is cleaner and better for all. I think the goal of being able to boot with empty /etc is a very worthy one, and is what we should get behind. For the compat cases use L and C, which should be good enough to provide compts with packages that really isn't up for this mode of operation.

I am not willing to accept that /etc is a lost cause, and we should just treat it the same as /usr with some ugly hacks around making some stuff of it overridable. We are in the business of doing this properly, hence let's do /etc right: focus on making it minimally populated, and writable if local configuration is necessary.

**Comment by boucman** [comment]
Oh, I totally agree with the goal... and we do our best (at least in the projects I have access to) to respect that convention.

However, if you boot with an empty /etc you will get the default configuration for all software. you (legitimately) need /etc to change the configuration. As I tried to explain, in the embedded world, that configuration is part of the image (mostly) there is no ansible overlay that is added after the fact, no docker image that are merged, and no admin to set his preferences.

I know the embedded use-case is hard to explain, because we break various concepts of normal distros (the biggest one being that systems can be individualized, and that there is someone to look into/fix things if they go wrong) 

OTA upgrades are the hardest part of embedded systems to understand. They need to be absolutely foolproof, always be able to rollback, and they must be able to do it on their own. This means (as an example) that we can't use package management in embedded systems. that's how it was done in the years 2010 and cause way more bricking of devices that what is publicly known. 
Nowdays, A/B is the only way to go because it provides safe rollback, but it also means that anything that is shared between A and B is a potential source of upgrade failures and must be minimized/audited in case of corruption.

This whole upgrade philosophy is what drives the whole partition management in the embedded use-case:
* anything on the A/B partition will be rewritten on upgrade (so it might as well be read-only, even if the hardware can support writing)
* there is a separate data partition which is used for locally generated, persistant data. It is used for /var.  the data partition can usually not be pre-populated in factory because it would consume a lot of bench-time. So it's usually formated on first boot (systemd-repart is awesome, btw, thx)

So /etc is problematic because it needs to contain a mixture of both...
* most of the sotware on the system (web servers, ntp servers...) are configured by upstream and never modified. So they should be part of the OTA upgrade. We want to avoid complex update mechanism and file copied around, because it's usually a source of robustness problem.
* some configuration is locally generated (I took the user's wifi credentials as an example earlier) and need to survive upgrade. So we store that inormation in the data partition and somehow (symlink, overlay, bind mount) make it available in /etc.

most of this works correctly, we tweak stuff before sysinit.target and use it after. The only file that is really problematic is... /etc/machine-id because it's used to early to intervene.

So yeah... I'm afraid the way forward for the embedded use-case is to go the initrd way. You suggested setting an environment variable... that could be a way forward if we set that in the boot loader somehow. But integration with boot loaders is a whole other can of worm since systemd is more and more UEFI dependant and there is basically no UEFI in the embedded world (the ARM UEFI is only really used on server chips. I still have to see it used in any real project)

So yeah, there are various ways around but they are all klunky and one of the strength of systemd is indeed to force a correct way to do stuff. I'd really love to have such a correct way for the embedded use-case and then push it to yocto/buildroot...

**Comment by poettering** [comment]
Did you read my suggestions above? i.e. use C/L tmpfiles.d/ lines to copy or symlink stuff from /usr/share/factory/ to /etc/ if it is missing there for the software that insists on having its config exclusively in /etc? Then, have immutable /usr and writable /etc and all is good. Since tmpfiles.d/ is run in very early boot all config is put into place there before most other userspace checks there. And you will have a mostly empty /etc that just contains the data you actually want to allow to change locally (machine ID) plus a skeleton of symlinks into /usr/share/factory/.

**Comment by boucman** [comment]
Ok, I get what you mean now... that could work except for the fact that the writable partition usually doesn't exist on first boot 

this is due to the way embedded devices are flashed in factory. we minimize the bench time as much as possible and writing to the flash through a JTAG is one of the longest step in the factory process. So we try to reduce the footprint, and formating a partition on first boot rather than before boot is a huge gain.

but I'm pretty sure we can find some workarounds for that too. The problem is more that the industry works more the way I describe than the way you would like it to work and I can't magically change that. 

But ok... at this point lets just close the bug since that's not the direction the project wants to go..

**Comment by poettering** [comment]
> Ok, I get what you mean now... that could work except for the fact that the writable partition usually doesn't exist on first boot

That's what "systemd-repart" is for, runs in the initrd, and adds in and formats partitions. including with encryption and stuff. The idea is that you only have to ship the A partition for /usr in your minimal image, and on first boot you get B for /usr added in, as well as the writable root partition formatted. "systemd-repart" is not supposed to be a workaround but the tool of choice for such things.

i don't actually think that embedded is that different from real systems in this regard. For example, for a ChromeOS-like device I'd also envision that your golden image is a minimized image consisting of an A partition for /usr, plus a verity of A, plus an EFI ESP and nothing more. Then on first boot, we possibly enlarge A (so that larger images can be added in eventually) and the verity of it, add in B + verity for B of the same size, then add an encrypted root partition plus a dm-integrity protected /home partition, plus maybe a swap partition. Thus /etc/ will come up empty and will be initialized with sysusers.d/ and tmpfiles.d/ stuff to carry what is really necessary but not more.

(there is some glue still missing to make this entirely clean: for example, we should define standardized GPT partition type uuids for /usr partitions — so far we only have them for the whole of root)

**Comment by gylstorffq** [comment]
If we have no `/etc/machine-id` on a firmware image and boot an error occurs and no `/etc/machine-id` will be created. As the default root mount of the kernel is RO and only later a RW mount occurs. This leads to first boot loop.

 If the `/etc/machine-id` is set to `uninitialized` a first boot will be executed. Afterwards `/etc/machine-id` contains the correct id. 

We have seen this behavior on Debian testing with systemd 251 and Fedora 36.

**Comment by Rahix** [comment]
@gylstorffq, if your `/etc/` starts read-only, the machine-id must be `uninitialized` in the firmware image so it can be over-mounted with the correct machine-id until `/etc/` becomes writable.  The problem is that systemd cannot bind-mount `/etc/machine-id` into thin air.  Once `/etc/` becomes writable, `systemd-machine-id-commit.service` will commit the machine-id into the filesystem persistently.

BTW, if you want to make use of `ConditionFirstBoot=`, you should also order your services before `first-boot-complete.target` to make sure they will have a chance to run at least once before the machine-id is committed and with that, the first boot marked as complete.

**Comment by fmoessbauer** [comment]
@Rahix Thanks for the clarification.

> (1) If /etc/machine-id does not exist, this is a first boot. During early boot, systemd will write "uninitialized\n" to this file and overmount a temporary file which contains the actual machine ID. Later (after first-boot-complete.target has been reached), the real machine ID will be written to disk.
> (2) If /etc/machine-id contains the string "uninitialized", a boot is also considered the first boot. The same mechanism as above applies.

IMO it would be good to point out this subtle difference in the documentation. Having a RO /etc during early boot is quite common, so (1) from the docs will almost never work.
